Use git url fragment to specify build reference and dir context.

calavera opened the pull request:
Fixes #7071

TODO:
- [x] Unit tests
- [x] Integration cli tests
- [x] Documentation

Signed-off-by: David Calavera david.calavera@gmail.com

calavera committed:
Use git url fragment to specify reference and dir context.

Signed-off-by: David Calavera <david.calavera@gmail.com>

jfrazelle commented:
cool!

icecrime commented:
I like the idea! @tiborvass @duglin please move to `2-needs-code-review` if it sounds reasonable to you too.

duglin commented:
design SGTM

duglin commented:
please squash the commits

duglin commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -637,13 +637,23 @@ an [*ADD*](/reference/builder/#add) instruction to reference a file in the
+in the branch `container`, as a context:
+
+  docker build https://github.com/docker/rootfs.git#container:docker
+
```
I think you may want to copy the table from https://github.com/docker/docker/issues/7071#issuecomment-50965433
I like being more explicit about the various options and that table explains it really well.

calavera replied:
sounds like a good idea.

moxiegirl replied:
We do support tables. You need to tick it up and do some tweaks to your headings.
@calavera ping me direct if you have specific questions on the markdown

--->

For example, as context, you need to run this command to use a directory called `docker` in the branch `container`:

  		$ docker build https://github.com/docker/rootfs.git#container:docker

The following table represents all the valid suffixes with their build contexts:

Build Syntax Suffix | Commit Used | Build Context Used
--------------------|-------------|-------------------
`myrepo.git` | `refs/heads/master` | `/`
`myrepo.git#mytag` | `refs/tags/mytag` | `/`
`myrepo.git#mybranch` | `refs/heads/mybranch` | `/`
`myrepo.git#abcdef` | `sha1 = abcdef` | `/`
`myrepo.git#:myfolder` | `refs/heads/master` | `/myfolder`
`myrepo.git#master:myfolder` | `refs/heads/master` | `/myfolder`
`myrepo.git#mytag:myfolder` | `refs/tags/mytag` | `/myfolder`
`myrepo.git#mybranch:myfolder` | `refs/heads/mybranch` | `/myfolder`
`myrepo.git#abcdef:myfolder` | `sha1 = abcdef` | `/myfolder`

Instead of specifying a context, you can pass a single Dockerfile in the
`URL` or pipe the file in via `STDIN`. To pipe a Dockerfile from `STDIN`:

	 $ docker build - < Dockerfile

If you use STDIN or specify a `URL`, the system places the contents into a
file called `Dockerfile`, and any `-f`, `--file` option is ignored. In this
scenario, there is no context.

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4221,6 +4221,35 @@ func (s *DockerSuite) TestBuildFromGIT(c *check.C) {
+	}
+	defer git.Close()
+
+	u := fmt.Sprintf("%s#master:docker", git.RepoURL)
```
Do we need test the other variants? like ...git#branch ...#hash    w/o :dir

duglin replied:
And :dir w/o #branch?

calavera replied:
we can, but those cases are already covered in the unit tests. This one is more to make sure that the end to end integration is still working.

duglin replied:
ok - haven't made it that far yet in the diff/review

calavera commented:
@eggtree I've borrowed your awesome table with the suffixes for the documentation, I hope you don't mind!

cpuguy83 commented:
Nice LGTM

duglin commented on pkg/urlutil/git.go:
```diff
@@ -8,11 +11,13 @@ var (
 		"git@",
 	}
+
+	urlPathWithFragmentSuffix = regexp.MustCompile(".git(:?#.+)?$")
```
I'm trying to parse this regexp.... I think this allows for:
```
.git:#abc
```
what that the intent?  If so, what does the ":" do and we should probably show it in the table.

calavera replied:
`:?` at the beginning of a regular expression means "don't save this in the group", because we don't really care about what's in the fragment in there.

duglin replied:
You sure about that?  When I play with it at: http://play.golang.org/p/X3_7h2Fq7z
I was expecting it to return false for:
```
regexp.MatchString(".git(:?#.+)?$", "foo.git:#abc")
```

calavera replied:
gosh, sorry, you're right. I meant `?:`. I've fixed the code and added tests for both cases.

duglin commented on utils/git.go:
```diff
@@ -19,20 +22,26 @@ func GitClone(remoteURL string) (string, error) {
+func cloneArgs(remoteURL *url.URL, root string) []string {
 	args := []string{"clone", "--recursive"}
-	shallow := true
+	shallow := len(remoteURL.Fragment) == 0
```
Why are we only doing a shallow clone when there's no fragment in the URL?  Is this a git restriction?

calavera replied:
yes. A shallow clone only gives you the last commit in the history, you cannot checkout a branch because your local repo doesn't know what that is. You need to run a `git fetch` if you want to do that, which is exactly the same as running a full clone.

So the logic is as follow:

1. Shallow clone by default because you don't want to do anything with the git history yet.
2. Full clone because we need to manipulate the repo to checkout a reference.

duglin commented on pkg/urlutil/git_test.go:
```diff
@@ -9,6 +9,7 @@ var (
 		"git@bitbucket.org:atlassianlabs/atlassian-docker.git",
 		"https://github.com/docker/docker.git",
 		"http://github.com/docker/docker.git",
+		"http://github.com/docker/docker.git#branch",
```
I think you support `.../docker.git#:dir` - meaning 'master' branch - so I think we should test it and add it to the table.

calavera replied:
`#.+` means `whatever after #` there is no need to test `:dir` because that already whatever.

duglin replied:
what I meant was to make sure your logic parsed it correctly, meaning it maps to master + a dir - I wasn't concerned about the go-lang matches() method.  But I see you added a test for it down below, so we're ok.

duglin commented on utils/git_test.go:
```diff
@@ -41,16 +44,126 @@ func TestCloneArgsDumbHttp(t *testing.T) {
+		{"master", "FROM scratch", false},
+		{":subdir", "", true},     // missing directory error
+		{":Dockerfile", "", true}, // not a directory error
+		{"master:subdir", "", true},
```
maybe add the "#:dir" test here (when 'dir' is valid) instead of where I commented above.

calavera replied:
I can do that :+1:

duglin commented on utils/git_test.go:
```diff
@@ -41,16 +44,126 @@ func TestCloneArgsDumbHttp(t *testing.T) {
+			t.Fatalf("Expected %v, was %v\n", c.exp, string(b))
+		}
+	}
+}
```
Is there a way to test your "shallow" logic?

calavera replied:
the shallow logic is already being tested in master.

calavera commented:
@duglin I've added those test cases for the urls and the directories that you mentioned. I've also added `:subdir` in the doc table as a valid option.

duglin commented:
LGTM

calavera commented:
Thanks @duglin. Moving it to docs review, since there are other maintainers that have already review the code too.

calavera commented:
Thanks for your comment @moxiegirl! I've made those changes :confetti_ball:

thaJeztah commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -637,13 +637,36 @@ an [*ADD*](/reference/builder/#add) instruction to reference a file in the
+`myrepo.git#mytag:myfolder` | `refs/tags/mytag` | `/myfolder`
+`myrepo.git#mybranch:myfolder` | `refs/heads/mybranch` | `/myfolder`
+`myrepo.git#abcdef:myfolder` | `sha1 = abcdef` | `/myfolder`
+
```
Just a remark / note for docs reviewers; we may need to make a decision at some point whether or not we want the Markdown tables formatted / aligned nicely in the source (outweighing the pros and cons).

thaJeztah commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -637,13 +637,36 @@ an [*ADD*](/reference/builder/#add) instruction to reference a file in the
+a branch, a tag or a commit sha. The second part represents a subdirectory
+inside the main root repository that will be used as a root context.
+
+For example, as context, you need to run this command to use a directory called `docker` in the branch `container`:
```
(I've missed @moxiegirl's comments, so this may be for a reason); I'd leave out the "as context, you need to", and simplify to:

> For example, run this command to use a directory called `docker` in the branch `container`:

thaJeztah commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -637,13 +637,36 @@ an [*ADD*](/reference/builder/#add) instruction to reference a file in the
 Local clones give you the ability to access private repositories using
 local user credentials, VPN's, and so forth.
 
+Git urls accept context configuration in their fragment section divided by a colon `:`.
```
Should there be a comma before "divided"? I'll leave that one for the experts :)

thaJeztah replied:
Also "separated" might be a better choice?

jamtur01 replied:
URLs

jamtur01 replied:
+1 to comma.

thaJeztah commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -637,13 +637,36 @@ an [*ADD*](/reference/builder/#add) instruction to reference a file in the
 
+Git urls accept context configuration in their fragment section divided by a colon `:`.
+The first part represents the reference that git will check out, this can be either
+a branch, a tag or a commit sha. The second part represents a subdirectory
```
Add a comma before `or`

jamtur01 replied:
SHA

thaJeztah commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -637,13 +637,36 @@ an [*ADD*](/reference/builder/#add) instruction to reference a file in the
+Git urls accept context configuration in their fragment section divided by a colon `:`.
+The first part represents the reference that git will check out, this can be either
+a branch, a tag or a commit sha. The second part represents a subdirectory
+inside the main root repository that will be used as a root context.
```
The repeated use of "root" is a bit confusing here. Perhaps just;

"The second part represents a subdirectory inside the repository that will be used as build context."

Or doesn't that cover the usage here?

thaJeztah commented:
Nice feature, thanks! 

Added some notes for docs, but you might want to have the others have their say 

/cc @moxiegirl @fredlf @jamtur01 @SvenDowideit

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -637,13 +637,36 @@ an [*ADD*](/reference/builder/#add) instruction to reference a file in the
 local user credentials, VPN's, and so forth.
 
+Git urls accept context configuration in their fragment section divided by a colon `:`.
+The first part represents the reference that git will check out, this can be either
```
Git

calavera commented:
Thanks for the review @thaJeztah, @jamtur01. I've addressed your comments.

thaJeztah commented:
Docs LGTM

moxiegirl commented:
LGTM
